File: Books/Math_Refresher/matlab/ch1.m
% Chapter 1

% Start of script
%-------------------------------------------------------------------------%
close all;                   	% close all figures
clearvars; clearvars -global;	% clear all variables
clc;                         	% clear the command terminal
format shortG;                 	% pick the most compact numeric display
format compact;                	% suppress excess blank lines


% functions
%-------------------------------------------------------------------------%
% f(x) = x^2
x = -2:0.1:2;
f_x = x.^2 - 1;
figure
subplot(121)
plot(x,f_x,'b.','Linewidth',3)
hold on
plot(x,f_x,'b')
hold off
subplot(122)
plot(f_x,x,'b.','Linewidth',3)
hold on
plot(f_x,x,'b')
hold off
saveas(gcf,'..\figures\functions','pdf')


% f(x) = 0.75x-6
x = -2:0.5:6;
f_x = 0.75.*x - 6;
figure
plot(x,f_x,'b.','Linewidth',3)
hold on
plot(x,f_x,'b')
plot(0,-6,'ro',...
     4,-3,'ro','Linewidth',2)
hold off
saveas(gcf,'..\figures\simple_line','pdf')


% f(x) = sin(x)./x
x = -6:0.1:6;
f_x = sin(x)./x;
figure
plot(x,f_x,'b.','Linewidth',3)
hold on
plot(x,f_x,'b')
hold off
saveas(gcf,'..\figures\graph13','pdf')


% f(x) = x^2 piecewise
x1 = 0:0.1:4;
f_x1 = -x1.^2+2;
x2 = -4:0.1:0;
f_x2 = x2.^2-2;
figure
plot(f_x1,x1,'b.',...
     f_x2,x2,'b.','Linewidth',3)
hold on
plot(f_x1,x1,'b',...
     f_x2,x2,'b')
plot(2,0,'b*',...
     -2,0,'b*','Linewidth',2)
hold off
saveas(gcf,'..\figures\graph15','pdf')


% f(x) = y = -3x + 2
x = -3:0.1:3;
f_x = -3.*x + 2;
figure
plot(x,f_x,'b.','Linewidth',3)
hold on
plot(x,f_x,'b')
hold off
saveas(gcf,'..\figures\ex1_20_1','pdf')


% f(x) = y = sqrt(2).*x - 3
x = -3:0.1:3;
f_x = sqrt(2).*x - 3;
figure
plot(x,f_x,'b.','Linewidth',3)
hold on
plot(x,f_x,'b')
hold off
saveas(gcf,'..\figures\ex1_20_2','pdf')



% Summation
%-------------------------------------------------------------------------%

% summation "by hand"
%------------------------------------%
n = 10; % set the stopping criteria
x = 1:1:n; % create the vector x
x_i = x(1); % initialize x_i with teh first index
% sum over the remaining indicies
for ii = 2:n
    x_i = x_i + x(ii);
end

result_1 = x_i

% compare to the built-in summation function
result_2 = sum(x)


% summation with a constant "by hand"
%------------------------------------%
n = 10; % set the stopping criteria
c = 2; % some constant
x = 1:1:n; % create the vector x
x_i = c*x(1); % initialize x_i with teh first index
% sum over the remaining indicies
for ii = 2:n
    x_i = x_i + c*x(ii);
end

result_1 = x_i


x = 1:1:n; % create the vector x
x_i = x(1); % initialize x_i with teh first index
% sum over the remaining indicies
for ii = 2:n
    x_i = x_i + x(ii);
end

result_2 = c*x_i


% Products
%-------------------------------------------------------------------------%

% products "by hand"
%------------------------------------%
n = 10; % set the stopping criteria
x = 1:1:n; % create the vector x
x_i = x(1); % initialize x_i with teh first index
% sum over the remaining indicies
for ii = 2:n
    x_i = x_i * x(ii);
end

result_1 = x_i

% compare to the built-in summation function
result_2 = prod(x)


% products with a constant "by hand"
%------------------------------------%
n = 10; % set the stopping criteria
c = 2; % some constant
x = 1:1:n; % create the vector x
x_i = c*x(1); % initialize x_i with teh first index
% sum over the remaining indicies
for ii = 2:n
    x_i = x_i * c*x(ii);
end

result_1 = x_i


x = 1:1:n; % create the vector x
x_i = x(1); % initialize x_i with teh first index
% sum over the remaining indicies
for ii = 2:n
    x_i = x_i * x(ii);
end

result_2 = c^n * x_i


% Factorials
%-------------------------------------------------------------------------%

x = 5;

% factorial
x_fac = x; % initialize
for ii = 1:x-1
    x_fac = x_fac * (x-ii);
end

result_1 = x_fac

% compare to the built-in function
result_2 = factorial(x)


% Modulo
%-------------------------------------------------------------------------%

result = mod(100,30)

result = mod(14,4)


% Log and exp
%-------------------------------------------------------------------------%

% examples
%------------------------------------%
x = 10;

% this will give us y = log_10(x)
result_1 = log10(x)

% this will give us 10^y = x
result_2 = 10^(log10(x))

% this will give us y = log_e(x)
result_1 = exp(log(x))

% this will give us e^y = x
result_2 = exp(log(x))

% properties of exponents
%------------------------------------%
a = 5;
x = 2;
y = 3;

% prop 1
result_1 = a^x * a^y
result_2 = a^(x+y)

% prop 2
result_1 = a^(-x)
result_2 = 1/(a^x)

% prop 3
result_1 = a^x / a^y
result_2 = a^(x-y)

% prop 4
result_1 = (a^x)^y
result_2 = a^(x*y)

% prop 5
result_1 = a^0


% properties of logs
%------------------------------------%
x = 2;
y = 3;

% prop 1
result_1 = log(x*y)
result_2 = log(x) + log(y)

% prop 2
result_1 = log(x^y)
result_2 = y*log(x)

% prop 3
result_1 = log(1/x)
result_2 = -log(x)

% prop 4
result_1 = log(x/y)
result_2 = log(x) - log(y)

% prop 5
result_1 = log(1)


% quadratic formula
%-------------------------------------------------------------------------%
% ax^2 + bx + c = 0

% example 2
% x^2 + 3x - 4 = 0
a = 1;
b = 3;
c = -4;

x_p = (-b + sqrt(b^2 - 4*a*c))/(2*a)
x_m = (-b - sqrt(b^2 - 4*a*c))/(2*a)


% Functions
%-------------------------------------------------------------------------%

% Paraboloid of One Sheet
%------------------------------------%
% Declare some constants.
a=2;
b=2;
c=1;
numPts = 30;
z_shift = 1; 
scale = 1;

% Define range of x and y axes.
x_0 = linspace(-2*a, 2*a, numPts);
y_0 = linspace(-2*b, 2*b, numPts);
[x, y] = meshgrid(x_0, y_0);

% Construct function.
% z = scale * sqrt(z_shift + x.^2 / a^2 + y.^2 / b^2);
z = scale * (z_shift + x.^2 / a^2 + y.^2 / b^2);
z = reshape(z, [], length(y_0));

figure;
surf(z);
hold on;
contour(z);
hold off;
grid on;
title('Paraboloid of One Sheet')
xlabel x
ylabel y
zlabel z


% Elliptical Paraboloid
%------------------------------------%
[r,theta] = meshgrid([0:0.05:5],[0:pi/50:2*pi]);

x = r.*cos(theta);
y = r.*sin(theta);
z = x.^2 + y.^2;

figure;
surf(x,y,z)
hold on;
contour(x,y,z);
hold off;
title('Elliptical Paraboloid')
xlabel x
ylabel y
zlabel z


% Hyperbolic Paraboloid
%------------------------------------%
x = [-10:.5:10];
y=[-10:.5:10];

[X, Y] = meshgrid(x,y);

Z = X.^2-Y.^2;

figure;
surf(X,Y,Z)
title('Hyperbolic Paraboloid')
xlabel y
ylabel x
zlabel z

File: Books/Math_Refresher/python/ch1.py
# ch1.py
import numpy as np
import matplotlib.pyplot as plt
from math import factorial, sqrt, floor
import math
from mpl_toolkits.mplot3d import Axes3D  # For 3D plotting

plt.close('all')

# ----------------------------
# f(x) = x^2 - 1
# ----------------------------
x = np.arange(-2, 2.1, 0.1)
f_x = x**2 - 1

plt.figure()
plt.subplot(1, 2, 1)
plt.plot(x, f_x, 'b.', linewidth=3)
plt.plot(x, f_x, 'b')
plt.subplot(1, 2, 2)
plt.plot(f_x, x, 'b.', linewidth=3)
plt.plot(f_x, x, 'b')
plt.savefig("../figures/functions.pdf")

# ----------------------------
# f(x) = 0.75*x - 6
# ----------------------------
x = np.arange(-2, 6.5, 0.5)
f_x = 0.75 * x - 6
plt.figure()
plt.plot(x, f_x, 'b.', linewidth=3)
plt.plot(x, f_x, 'b')
plt.plot(0, -6, 'ro', markersize=8)
plt.plot(4, -3, 'ro', markersize=8)
plt.savefig("../figures/simple_line.pdf")

# ----------------------------
# f(x) = sin(x)/x
# ----------------------------
x = np.arange(-6, 6.1, 0.1)
f_x = np.sin(x) / x
f_x[np.isnan(f_x)] = 1  # Handle division by zero at x = 0
plt.figure()
plt.plot(x, f_x, 'b.', linewidth=3)
plt.plot(x, f_x, 'b')
plt.savefig("../figures/graph13.pdf")

# ----------------------------
# Piecewise function:
# f(x) = -x^2+2 for x>=0 and f(x) = x^2-2 for x<0
# ----------------------------
x1 = np.arange(0, 4.1, 0.1)
f_x1 = -x1**2 + 2
x2 = np.arange(-4, 0.1, 0.1)
f_x2 = x2**2 - 2
plt.figure()
plt.plot(f_x1, x1, 'b.', markersize=8, linewidth=3, label='x>=0')
plt.plot(f_x2, x2, 'b.', markersize=8, linewidth=3, label='x<0')
plt.plot(f_x1, x1, 'b', linewidth=1)
plt.plot(f_x2, x2, 'b', linewidth=1)
plt.plot(2, 0, 'b*', markersize=10)
plt.plot(-2, 0, 'b*', markersize=10)
plt.savefig("../figures/graph15.pdf")

# ----------------------------
# f(x) = -3*x + 2
# ----------------------------
x = np.arange(-3, 3.1, 0.1)
f_x = -3 * x + 2
plt.figure()
plt.plot(x, f_x, 'b.', linewidth=3)
plt.plot(x, f_x, 'b')
plt.savefig("../figures/ex1_20_1.pdf")

# ----------------------------
# f(x) = sqrt(2)*x - 3
# ----------------------------
x = np.arange(-3, 3.1, 0.1)
f_x = np.sqrt(2) * x - 3
plt.figure()
plt.plot(x, f_x, 'b.', linewidth=3)
plt.plot(x, f_x, 'b')
plt.savefig("../figures/ex1_20_2.pdf")

# ----------------------------
# Summation "by hand"
# ----------------------------
n = 10
x_vec = np.arange(1, n+1)
x_i = x_vec[0]
for ii in range(1, n):
    x_i += x_vec[ii]
result_1 = x_i
result_2 = np.sum(x_vec)
print("Summation by hand:", result_1, "Built-in sum:", result_2)

# ----------------------------
# Summation with a constant "by hand"
# ----------------------------
n = 10
c = 2
x_vec = np.arange(1, n+1)
x_i = c * x_vec[0]
for ii in range(1, n):
    x_i += c * x_vec[ii]
result_1 = x_i

x_vec = np.arange(1, n+1)
x_i = x_vec[0]
for ii in range(1, n):
    x_i += x_vec[ii]
result_2 = c * x_i
print("Summation with constant by hand:", result_1, "and", result_2)

# ----------------------------
# Products "by hand"
# ----------------------------
n = 10
x_vec = np.arange(1, n+1)
x_i = x_vec[0]
for ii in range(1, n):
    x_i *= x_vec[ii]
result_1 = x_i
result_2 = np.prod(x_vec)
print("Product by hand:", result_1, "Built-in prod:", result_2)

# ----------------------------
# Products with a constant "by hand"
# ----------------------------
n = 10
c = 2
x_vec = np.arange(1, n+1)
x_i = c * x_vec[0]
for ii in range(1, n):
    x_i *= c * x_vec[ii]
result_1 = x_i

x_vec = np.arange(1, n+1)
x_i = x_vec[0]
for ii in range(1, n):
    x_i *= x_vec[ii]
result_2 = (c**n) * x_i
print("Product with constant by hand:", result_1, "and", result_2)

# ----------------------------
# Factorials
# ----------------------------
x_val = 5
x_fac = x_val
for ii in range(1, x_val):
    x_fac *= (x_val - ii)
result_1 = x_fac
result_2 = factorial(x_val)
print("Factorial by hand:", result_1, "Built-in factorial:", result_2)

# ----------------------------
# Modulo
# ----------------------------
result = 100 % 30
print("100 mod 30 =", result)
result = 14 % 4
print("14 mod 4 =", result)

# ----------------------------
# Log and exp examples
# ----------------------------
x_val = 10
result_1 = np.log10(x_val)
result_2 = 10**(np.log10(x_val))
print("log10(10) =", result_1, "10^(log10(10)) =", result_2)
result_1 = np.exp(np.log(x_val))
print("exp(log(10)) =", result_1)

# ----------------------------
# Properties of exponents and logs
# ----------------------------
a = 5; x_val = 2; y_val = 3
print("a^x * a^y =", a**x_val * a**y_val, "and a^(x+y) =", a**(x_val+y_val))
print("a^(-x) =", a**(-x_val), "and 1/a^x =", 1/(a**x_val))
print("a^x / a^y =", a**x_val / a**y_val, "and a^(x-y) =", a**(x_val-y_val))
print("(a^x)^y =", (a**x_val)**y_val, "and a^(x*y) =", a**(x_val*y_val))
print("a^0 =", a**0)
print("log(x*y) =", np.log(x_val*y_val), "and log(x)+log(y) =", np.log(x_val)+np.log(y_val))
print("log(x^y) =", np.log(x_val**y_val), "and y*log(x) =", y_val*np.log(x_val))
print("log(1/x) =", np.log(1/x_val), "and -log(x) =", -np.log(x_val))
print("log(x/y) =", np.log(x_val/y_val), "and log(x)-log(y) =", np.log(x_val)-np.log(y_val))
print("log(1) =", np.log(1))

# ----------------------------
# Quadratic formula: Solve x^2 + 3x - 4 = 0
# ----------------------------
a_coef = 1; b_coef = 3; c_coef = -4
x_p = (-b_coef + np.sqrt(b_coef**2 - 4*a_coef*c_coef)) / (2*a_coef)
x_m = (-b_coef - np.sqrt(b_coef**2 - 4*a_coef*c_coef)) / (2*a_coef)
print("Quadratic formula solutions:", x_p, x_m)

# ----------------------------
# 3D Surface Plots: Functions
# ----------------------------

# Paraboloid of One Sheet
a_const = 2; b_const = 2; numPts = 30; z_shift = 1; scale = 1
x_0 = np.linspace(-2*a_const, 2*a_const, numPts)
y_0 = np.linspace(-2*b_const, 2*b_const, numPts)
X, Y = np.meshgrid(x_0, y_0)
Z = scale * (z_shift + X**2 / a_const**2 + Y**2 / b_const**2)
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
ax.plot_surface(X, Y, Z, cmap='viridis')
ax.contour(X, Y, Z, zdir='z', offset=np.min(Z), cmap='viridis')
ax.set_title('Paraboloid of One Sheet')
ax.set_xlabel('x'); ax.set_ylabel('y'); ax.set_zlabel('z')
plt.savefig("../figures/paraboloid_one_sheet.pdf")

# Elliptical Paraboloid
r, theta = np.meshgrid(np.arange(0, 5.05, 0.05), np.arange(0, 2*np.pi+0.01, np.pi/50))
x_val = r * np.cos(theta)
y_val = r * np.sin(theta)
z_val = x_val**2 + y_val**2
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
ax.plot_surface(x_val, y_val, z_val, cmap='viridis')
ax.contour(x_val, y_val, z_val, zdir='z', offset=np.min(z_val), cmap='viridis')
ax.set_title('Elliptical Paraboloid')
ax.set_xlabel('x'); ax.set_ylabel('y'); ax.set_zlabel('z')
plt.savefig("../figures/elliptical_paraboloid.pdf")

# Hyperbolic Paraboloid
x_line = np.arange(-10, 10.5, 0.5)
y_line = np.arange(-10, 10.5, 0.5)
X, Y = np.meshgrid(x_line, y_line)
Z = X**2 - Y**2
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
ax.plot_surface(X, Y, Z, cmap='viridis')
ax.set_title('Hyperbolic Paraboloid')
ax.set_xlabel('y'); ax.set_ylabel('x'); ax.set_zlabel('z')
plt.savefig("../figures/hyperbolic_paraboloid.pdf")

plt.show()
